### ISSUE TITLE
azure-spring-boot breaks actuator/env

### ISSUE THREAD

**Issue by yossaa** [Microsoft/azure-spring-boot#643]
## Environment
- Spring boot starter: 
  - key vault spring boot starter
  - servicebus (Vcap)
  
- OS Type: Windows/Linux
- Java version: 
  - 1.8

## Summary
I have integrated Azure Keyvault in our application and then noticed that actuator/env Endpoint stopped working. It throws :

> Type definition error: [simple type, class java.util.Random]; nested exception is com.fasterxml.jackson.databind.exc.InvalidDefinitionException: No serializer found for class java.util.Random and no properties discovered to create BeanSerializer (to avoid exception, disable SerializationFeature.FAIL_ON_EMPTY_BEANS) (through reference chain: org.springframework.boot.actuate.env.EnvironmentEndpoint$EnvironmentDescriptor["propertySources"]->java.util.ArrayList[9]->org.springframework.boot.actuate.env.EnvironmentEndpoint$PropertySourceDescriptor["properties"]->java.util.LinkedHashMap["azure.storage.result"]->org.springframework.boot.actuate.env.EnvironmentEndpoint$PropertyValueDescriptor["value"]->com.microsoft.azure.spring.cloundfoundry.environment.VcapResult["confEnv"]->org.springframework.core.env.StandardEnvironment["propertySources"]->org.springframework.boot.context.properties.source.ConfigurationPropertySourcesPropertySource$$EnhancerBySpringCGLIB$$f58669f1["source"]->org.springframework.boot.context.properties.source.SpringConfigurationPropertySource["underlyingSource"]->com.ulisesbocchio.jasyptspringboot.wrapper.EncryptablePropertySourceWrapper["source"])

The problem is the VcapResult class which lands in the EnvironmentDescriptor and it contains confEnv property with all PropertySources, including RandomValuePropertySource:

![image](https://user-images.githubusercontent.com/14076521/55066624-7f19a800-507e-11e9-96a8-d135ab2acccf.png)

![image](https://user-images.githubusercontent.com/14076521/55064720-bd14cd00-507a-11e9-9d19-a88c151250be.png)

My workaround for now is to remove those PropertySources in my own PostProcessor:

```
public class DisableAzureVcapPostProcessor implements EnvironmentPostProcessor {
  private static final Set<String> keyNames = ImmutableSet.of(
      "azure.storage.result",
      "azure.documentdb.result",
      "azure.servicebus.result");

  public void postProcessEnvironment(ConfigurableEnvironment confEnv, SpringApplication app) {
    PropertySource<?> source = confEnv.getPropertySources().get("defaultProperties");
    if (source != null && source instanceof MapPropertySource) {
      keyNames.forEach( key -> ((MapPropertySource)source).getSource().put(key, ""));
    }
  }
}
```

registered in spring.factories:
```
org.springframework.boot.env.EnvironmentPostProcessor=de.vattenfall.is.heat.base.azure.DisableAzureVcapPostProcessor
```

**Comment by Incarnation-p-lee** [comment]
@yossaa
Thanks for issue this. Could you please provide more information about this issue, like starter version, spring boot version, minimal code for reproducing ? So we can try to root cause the problem.

**Comment by yossaa** [comment]
sure. You need no code to reproduce it, just add azure-keyvault-secrets-spring-boot-starter to your app.

src/main/java/test
```
package test;

import java.util.Properties;

import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;

@SpringBootApplication
public class Test {
  public static void main(String[] args) {
    SpringApplication application = new SpringApplication(Test.class);

    Properties properties = new Properties();
    properties.put("server.port", 9999);
    properties.put("management.endpoints.web.exposure.include", "*");

    application.setDefaultProperties(properties);
    application.run(args);
  }

}
```

pom.xml
```
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="http://maven.apache.org/POM/4.0.0"
         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>

    <groupId>test</groupId>
    <artifactId>azure-test</artifactId>
    <version>0.0.1</version>
    <packaging>jar</packaging>

    <parent>
        <groupId>org.springframework.boot</groupId>
        <artifactId>spring-boot-starter-parent</artifactId>
        <version>2.1.3.RELEASE</version>
    </parent>
    <properties>
        <java.version>1.8</java.version>
        <lib.version.spring-cloud>Greenwich.RELEASE</lib.version.spring-cloud>
        <lib.version.spring>2.1.3.RELEASE</lib.version.spring>
    </properties>

    <dependencies>
        <dependency>
            <groupId>org.springframework.cloud</groupId>
            <artifactId>spring-cloud-dependencies</artifactId>
            <version>${lib.version.spring-cloud}</version>
            <type>pom</type>
            <scope>compile</scope>
        </dependency>
        <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-starter-web</artifactId>
        </dependency>
        <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-starter-actuator</artifactId>
        </dependency>
        <dependency>
            <groupId>com.microsoft.azure</groupId>
            <artifactId>azure-keyvault-secrets-spring-boot-starter</artifactId>
            <version>2.1.6</version>
        </dependency>
    </dependencies>
    <build>
        <plugins>
            <plugin>
                <groupId>org.springframework.boot</groupId>
                <artifactId>spring-boot-maven-plugin</artifactId>
                <version>2.1.3.RELEASE</version>
            </plugin>
        </plugins>
    </build>
</project>
```

```
mvn package
java -jar target/azure-test-0.0.1.jar
```

Open: http://localhost:9999/actuator/env
Check exception on console.

In this case I get:
```
Cannot render error page for request [/actuator/env] and exception [Type definition error: [simple type, class java.util.logging.ErrorManager]; nested exception is com.fasterxml.jackson.databind.exc.InvalidDefinitionException: No serializer found for class java.util.logging.ErrorManager
```

Classes causing serialization problems are random of course. You put in the Environment objects, which shouldn't be there, because they are not serializable.

**Comment by yossaa** [comment]
btw. in your current design all azure* starters are loading all components. In this case I wanted to use keyvault, and your lib is trying to autoconfigure all other components (storage, cosmos, systembus).

**Comment by Incarnation-p-lee** [comment]
@yossa
I am afraid I cannot reproduce you issue from my side, with you provided code. Spring boot 2.1.x may change the endpoint after /actuator, you may need to double confirm if `/actuator/env` exist or not.

About the design, keyvault starter as post processor will active unconditionally. But for rest starters, it only active when related properties exist, which is not the same as loading all components anywhere. That is what `auto-configuration` do from my perspective, may be there is some gap on `auto-configuration` from you side.

```
<62/2404><>panli^~/Desktop/demo@<09:44:48 Fri Mar 29>[0]*1*[5s]"java -jar target/demo-0.0.1-SNAPSHOT.jar &"
>>
  .   ____          _            __ _ _
 /\\ / ___'_ __ _ _(_)_ __  __ _ \ \ \ \
( ( )\___ | '_ | '_| | '_ \/ _` | \ \ \ \
 \\/  ___)| |_)| | | | | || (_| |  ) ) ) )
  '  |____| .__|_| |_|_| |_\__, | / / / /
 =========|_|==============|___/=/_/_/_/
 :: Spring Boot ::        (v2.1.3.RELEASE)

2019-03-29 09:44:49.472  INFO 1272 --- [           main] com.example.demo.DemoApplication         : Starting DemoApplication v0.0.1-SNAPSHOT on panli with PID 1272 (C:\Users\panli\Desktop\demo\target\demo-0.0.1-SNAPSHOT.jar started by panli in C:\Users\panli\Desktop\demo)
2019-03-29 09:44:49.474  INFO 1272 --- [           main] com.example.demo.DemoApplication         : No active profile set, falling back to default profiles: default
2019-03-29 09:44:50.973  INFO 1272 --- [           main] o.s.b.w.embedded.tomcat.TomcatWebServer  : Tomcat initialized with port(s): 8080 (http)
2019-03-29 09:44:51.007  INFO 1272 --- [           main] o.apache.catalina.core.StandardService   : Starting service [Tomcat]
2019-03-29 09:44:51.007  INFO 1272 --- [           main] org.apache.catalina.core.StandardEngine  : Starting Servlet engine: [Apache Tomcat/9.0.16]
2019-03-29 09:44:51.018  INFO 1272 --- [           main] o.a.catalina.core.AprLifecycleListener   : The APR based Apache Tomcat Native library which allows optimal performance in production environments was not found on the java.library.path: [C:\Program Files (x86)\Common Files\Oracle\Java\javapath;C:\WINDOWS\Sun\Java\bin;C:\WINDOWS\system32;C:\WINDOWS;C:\Users\panli\bin;C:\Program Files\Git\mingw64\bin;C:\Program Files\Git\usr\local\bin;C:\Program Files\Git\usr\bin;C:\Program Files\Git\usr\bin;C:\Program Files\Git\cmd;C:\Program Files\Git\usr\bin;C:\Program Files\ConEmu\ConEmu\Scripts;C:\Program Files\ConEmu;C:\Program Files\ConEmu\ConEmu;C:\Program Files (x86)\Common Files\Oracle\Java\javapath;C:\Program Files\Docker\Docker\resources\bin;C:\Program Files (x86)\Microsoft SDKs\Azure\CLI2\wbin;C:\ProgramData\Oracle\Java\javapath;C:\Program Files\Common Files\microsoft shared\Microsoft Online Services;C:\Program Files (x86)\Common Files\microsoft shared\Microsoft Online Services;C:\Windows\System32;C:\Windows;C:\Windows\System32\wbem;C:\Windows\System32\WindowsPowerShell\v1.0;C:\Windows\System32\OpenSSH;C:\Program Files (x86)\GnuPG\bin;C:\Program Files\Microsoft VS Code\bin;C:\Go\bin;C:\Program Files\dotnet;C:\Program Files\Microsoft SQL Server\130\Tools\Binn;C:\Program Files\Git\cmd;C:\Users\panli\AppData\Local\Microsoft\WindowsApps;C:\Users\panli\Desktop\Applications\apache-maven-3.5.2\bin;C:\Program Files\Microsoft VS Code\bin;C:\Gradle\gradle-4.5.1\bin;C:\Program Files\Java\jdk1.8.0_171\bin;C:\Users\panli\Desktop\Applications\apache-gremlin-console-3.1.0-incubating\bin;C:\Users\panli\AppData\Local\Programs\Python\Python36-32;C:\Users\panli\AppData\Local\Programs\Fiddler;C:\Users\panli\AppData\Local\Microsoft\WindowsApps;C:\Program Files\Microsoft SDKs\Azure\Azure Dev Spaces CLI (Preview);C:\Users\panli\go\bin;C:\Program Files\Git\usr\bin\vendor_perl;C:\Program Files\Git\usr\bin\core_perl;.]
2019-03-29 09:44:51.187  INFO 1272 --- [           main] o.a.c.c.C.[Tomcat].[localhost].[/]       : Initializing Spring embedded WebApplicationContext
2019-03-29 09:44:51.187  INFO 1272 --- [           main] o.s.web.context.ContextLoader            : Root WebApplicationContext: initialization completed in 1671 ms
2019-03-29 09:44:51.650  INFO 1272 --- [           main] o.s.s.concurrent.ThreadPoolTaskExecutor  : Initializing ExecutorService 'applicationTaskExecutor'
2019-03-29 09:44:51.960  INFO 1272 --- [           main] o.s.b.a.e.web.EndpointLinksResolver      : Exposing 2 endpoint(s) beneath base path '/actuator'
2019-03-29 09:44:52.030  INFO 1272 --- [           main] o.s.b.w.embedded.tomcat.TomcatWebServer  : Tomcat started on port(s): 8080 (http) with context path ''
2019-03-29 09:44:52.033  INFO 1272 --- [           main] com.example.demo.DemoApplication         : Started DemoApplication in 2.919 seconds (JVM running for 3.328)

<62/2404><>panli^~/Desktop/demo@<09:44:54 Fri Mar 29>[0]*1*[6s]"java -jar target/demo-0.0.1-SNAPSHOT.jar &"
>> curl localhost:8080/actuator/
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     02019-03-29 09:45:10.005  INFO 1272 --- [nio-8080-exec-1] o.a.c.c.C.[Tomcat].[localhost].[/]       : Initializing Spring DispatcherServlet 'dispatcherServlet'
2019-03-29 09:45:10.006  INFO 1272 --- [nio-8080-exec-1] o.s.web.servlet.DispatcherServlet        : Initializing Servlet 'dispatcherServlet'
2019-03-29 09:45:10.014  INFO 1272 --- [nio-8080-exec-1] o.s.web.servlet.DispatcherServlet        : Completed initialization in 8 ms
100   441    0   441    0     0   3242      0 --:--:-- --:--:-- --:--:--  3242{"_links":{"self":{"href":"http://localhost:8080/actuator","templated":false},"health-component":{"href":"http://localhost:8080/actuator/health/{component}","templated":true},"health-component-instance":{"href":"http://localhost:8080/actuator/health/{component}/{instance}","templated":true},"health":{"href":"http://localhost:8080/actuator/health","templated":false},"info":{"href":"http://localhost:8080/actuator/info","templated":false}}}
```

**Comment by yossaa** [comment]
In spring boot 2 you have to add a property:
management.endpoints.web.exposure.include=*

Otherwise you don't see /env

**Comment by Incarnation-p-lee** [comment]
@yossaa 
Thanks for pointing out this. It should be cloud foundry properties which cannot be converted to Json. We will prepare one PR for this.

**Comment by berndgoetz** [comment]
I can reproduce this with my test project at https://github.com/berndgoetz/spring-sec-aad-sample.

**Comment by Incarnation-p-lee** [comment]
Could you please have a try with snapshot ? The fixing PR has merged already.

**Comment by yossaa** [comment]
I have just tried with 2.1.7-SNAPSHOT and still getting this exception:
![image](https://user-images.githubusercontent.com/14076521/56806878-807cf280-682d-11e9-9f10-b2e28eadc1d0.png)

**Comment by Incarnation-p-lee** [comment]
Could you please have a try with snapshot ? Seems I missed Getter and Setter for VcapResult. Sorry for inconvenient.

**Comment by yossaa** [comment]
it works now

**Comment by xscript** [comment]
Closing it now.

**Comment by berndgoetz** [comment]
Is it possible that this fix has not been released yet? I only see 2.1.6 released. By when will there be a 2.1.7 release?

**Comment by xscript** [comment]
2.1.7 is on the way. Add @jialindai for more details.

**Comment by tjuchniewicz** [comment]
Can we apply this fix in 2.0.x?

**Comment by neuqlz** [comment]
@tjuchniewicz thanks for contact us, because we need to take efforts on version 2.2.0, so we won't apply this fix in 2.0.x. You can use 2.1.7 or 2.1.8.SNAPSHOT, sorry for the inconvenience to you.
